olp-cpp-sdk-dataservice-read/src/CatalogClientImpl.cpp

```diff
@@ -124,6 +125,47 @@ CancellableFuture<CatalogVersionResponse> CatalogClientImpl::GetLatestVersion(
   return client::CancellableFuture<CatalogVersionResponse>(
       std::move(cancel_token), std::move(promise));
 }
+
+client::CancellationToken CatalogClientImpl::ListVersions(
+    VersionsRequest request, VersionsResponseCallback callback) {
+  if (request.GetFetchOption() == CacheWithUpdate) {
+    auto task = [](client::CancellationContext) -> VersionsResponse {
+      return {{client::ErrorCode::InvalidArgument,
+               "CacheWithUpdate option can not be used for versioned catalog"}};
+    };
+    return AddTask(task_scheduler_, pending_requests_, std::move(task),
+                   std::move(callback));
+  }
```

Comment:
I think, it can be merged into `schedule_get_versions_list`. This will simplify the code since no need to create additional lamdas.